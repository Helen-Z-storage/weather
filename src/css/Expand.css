
.detailWeather {
    margin: 0.5em 0px;
  }

.expand {
  position: absolute;
  background-color: white;
  color: #A0A0A0;
  font-size: 4em;
  border: hidden;
  float: right;
  top: 0px;
  right: 0px;
}

.expand:focus {
  color: black;
}