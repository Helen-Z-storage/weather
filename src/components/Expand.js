import React from 'react';
import '../css/Expand.css';


function Expand(props) {
  const {id, currCity, expand, handleExpand} = props;


    
    let detailWeatherList = <li></li>;
    let currWeatherList = <li></li>;

    const currWeather = currCity.curr_forecast;
    const detailWeathers = currCity.detail_forecast;
    
    // expanded 
    detailWeatherList = 
    <li key={"forcast_weather"}>
      <table>
        <thead>
              <tr>
                  <td>{"Date and time"}</td>
                  <td>{"Forcast Weather"}</td>
                  <td>{"Forcast Temperature"}</td>
                  <td>{"Forcast Body Temperature"}</td>
                  <td>{"Forcast Max Temperature"}</td>
                  <td>{"Forcast Min Temperature"}</td>
                </tr>
            </thead>
            <tbody>
        {detailWeathers.map(
      (weatherObj, i) => 
              <tr key={i}>
                <td>{`${weatherObj.timestamp}`}</td>
                <td><img className='expand_img' alt={`${currCity.city} ${weatherObj.timestamp} weather icon`} src={weatherObj.icon} /></td>
                <td>{`${weatherObj.temp} C`}</td>
                <td>{`${weatherObj.feels_like} C`}</td>
                <td>{`${weatherObj.temp_min} C`}</td>
                <td>{`${weatherObj.temp_max} C`}</td>
              </tr>)}
              
            </tbody>
          </table>
          </li>

      currWeatherList = 
      <li key={"curr_weather"}>
            <table>
              <thead>
                <tr>
                    <td>{"Date and time"}</td>
                    <td>{"Current Weather"}</td>
                    <td>{"Current Temperature"}</td>
                    <td>{"Current Body Temperature"}</td>
                    <td>{"Current Max Temperature"}</td>
                    <td>{"Current Min Temperature"}</td>
                  </tr>
              </thead>
              <tbody>
                <tr>
                  <td>{`${currWeather.timestamp}`}</td>
                  <td><img className='expand_img' alt={`${currCity.city} ${currWeather.timestamp} weather icon`} src={currWeather.icon} /></td>
                  <td>{`${currWeather.temp} C`}</td>
                  <td>{`${currWeather.feels_like} C`}</td>
                  <td>{`${currWeather.temp_min} C`}</td>
                  <td>{`${currWeather.temp_max} C`}</td>
                </tr>
              </tbody>
            </table>
            </li>;
    return (
        <ul className='detailWeather'>
            <li><button className="expand" onClick={() => handleExpand(id)}>
                {expand? "-": "+"}</button></li>
            {expand? detailWeatherList: currWeatherList}
        </ul>
        );
}

export default Expand;