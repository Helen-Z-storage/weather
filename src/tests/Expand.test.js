import React from 'react'
import { create } from "react-test-renderer";
import { render, screen } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect' // expect function
import userEvent from '@testing-library/user-event'; 

import Expand from '../components/Expand';

// required prop data of create component
const user = userEvent.setup();

const id = 3;
const currCity = {
    "country": "Algeria",
    "city": "Algiers",
    "lat": 36.7753606,
    "lon": 3.0601882,
    "tags": ",",
    "icon": "https://openweathermap.org/img/wn/10d@2x.png",
    "curr_forecast": {
        "timestamp": "2023-01-18 09:00:00",
        "temp": 12.01,
        "feels_like": 11.12,
        "temp_min": 9.79,
        "temp_max": 12.01,
        "icon": "https://openweathermap.org/img/wn/10d@2x.png"
    },
    "detail_forecast": [
        {
            "timestamp": "2023-01-18 09:00:00",
            "temp": 12.01,
            "feels_like": 11.12,
            "temp_min": 9.79,
            "temp_max": 12.01,
            "icon": "https://openweathermap.org/img/wn/10d@2x.png"
        },
        {
            "timestamp": "2023-01-18 12:00:00",
            "temp": 11.35,
            "feels_like": 10.42,
            "temp_min": 10.47,
            "temp_max": 11.35,
            "icon": "https://openweathermap.org/img/wn/10d@2x.png"
        },
        {
            "timestamp": "2023-01-18 15:00:00",
            "temp": 11.8,
            "feels_like": 10.52,
            "temp_min": 11.8,
            "temp_max": 11.8,
            "icon": "https://openweathermap.org/img/wn/10d@2x.png"
        },
        {
            "timestamp": "2023-01-18 18:00:00",
            "temp": 11.07,
            "feels_like": 9.8,
            "temp_min": 11.07,
            "temp_max": 11.07,
            "icon": "https://openweathermap.org/img/wn/10n@2x.png"
        },
        {
            "timestamp": "2023-01-18 21:00:00",
            "temp": 11.57,
            "feels_like": 10.24,
            "temp_min": 11.57,
            "temp_max": 11.57,
            "icon": "https://openweathermap.org/img/wn/10n@2x.png"
        },
        {
            "timestamp": "2023-01-19 00:00:00",
            "temp": 11.42,
            "feels_like": 9.95,
            "temp_min": 11.42,
            "temp_max": 11.42,
            "icon": "https://openweathermap.org/img/wn/10n@2x.png"
        },
        {
            "timestamp": "2023-01-19 03:00:00",
            "temp": 10.93,
            "feels_like": 9.54,
            "temp_min": 10.93,
            "temp_max": 10.93,
            "icon": "https://openweathermap.org/img/wn/10n@2x.png"
        },
        {
            "timestamp": "2023-01-19 06:00:00",
            "temp": 10.43,
            "feels_like": 9.12,
            "temp_min": 10.43,
            "temp_max": 10.43,
            "icon": "https://openweathermap.org/img/wn/10n@2x.png"
        }
    ]
};
const expandT = true;
const expandF = false;
const handleExpand = jest.fn();

describe('testing Expand component', () => {

    test("Matches the snapshot: shows expanded Expand component", () => {
        // create component
        const expandData = create(
            <Expand id={id} currCity={currCity}
                    expand={expandT} handleExpand={handleExpand} />
        );

        // expecting output
        expect(expandData.toJSON()).toMatchSnapshot();
    });

    test("Matches the snapshot: shows non-expanded Expand component", () => {
        // create component
        const expandData = create(
            <Expand id={id} currCity={currCity}
                    expand={expandF} handleExpand={handleExpand} />
        );

        // expecting output
        expect(expandData.toJSON()).toMatchSnapshot();
    });

    test("Matches the snapshot: shows empty city", () => {
        // create component
        const expandData = create(
            <Expand id={id} currCity={{}}
                    expand={expandF} handleExpand={handleExpand} />
        );

        // expecting output
        expect(expandData.toJSON()).toMatchSnapshot();
    });

    test("shows clicked button working", async () => {
        // render component
        render(
            <Expand id={id} currCity={currCity}
                    expand={expandF} handleExpand={handleExpand} />
        );

        // expecting output
        const button = screen.getByText('+');
        await user.click(button); 
        expect(handleExpand).toBeCalledTimes(1);
        expect(handleExpand.mock.calls[0][0]).toEqual(id);
    });
});