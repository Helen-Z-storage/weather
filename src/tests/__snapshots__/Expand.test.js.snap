// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Expand component Matches the snapshot: shows empty city 1`] = `
<ul
  className="detailWeather"
>
  <li>
    <button
      className="expand"
      onClick={[Function]}
    >
      +
    </button>
  </li>
  <li />
</ul>
`;

exports[`testing Expand component Matches the snapshot: shows expanded Expand component 1`] = `
<ul
  className="detailWeather"
>
  <li>
    <button
      className="expand"
      onClick={[Function]}
    >
      -
    </button>
  </li>
  <li>
    <table>
      <thead>
        <tr>
          <td>
            Date & time
          </td>
          <td>
            Forcast Weather
          </td>
          <td>
            Forcast Temp
          </td>
          <td>
            Forcast Body Temp
          </td>
          <td>
            Forcast Max Temp
          </td>
          <td>
            Forcast Min Temp
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 09:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10d@2x.png"
            />
          </td>
          <td>
            12.01 °C
          </td>
          <td>
            11.12 °C
          </td>
          <td>
            9.79 °C
          </td>
          <td>
            12.01 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 12:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10d@2x.png"
            />
          </td>
          <td>
            11.35 °C
          </td>
          <td>
            10.42 °C
          </td>
          <td>
            10.47 °C
          </td>
          <td>
            11.35 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 15:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10d@2x.png"
            />
          </td>
          <td>
            11.8 °C
          </td>
          <td>
            10.52 °C
          </td>
          <td>
            11.8 °C
          </td>
          <td>
            11.8 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 18:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10n@2x.png"
            />
          </td>
          <td>
            11.07 °C
          </td>
          <td>
            9.8 °C
          </td>
          <td>
            11.07 °C
          </td>
          <td>
            11.07 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 21:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10n@2x.png"
            />
          </td>
          <td>
            11.57 °C
          </td>
          <td>
            10.24 °C
          </td>
          <td>
            11.57 °C
          </td>
          <td>
            11.57 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-19 00:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10n@2x.png"
            />
          </td>
          <td>
            11.42 °C
          </td>
          <td>
            9.95 °C
          </td>
          <td>
            11.42 °C
          </td>
          <td>
            11.42 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-19 03:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10n@2x.png"
            />
          </td>
          <td>
            10.93 °C
          </td>
          <td>
            9.54 °C
          </td>
          <td>
            10.93 °C
          </td>
          <td>
            10.93 °C
          </td>
        </tr>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-19 06:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10n@2x.png"
            />
          </td>
          <td>
            10.43 °C
          </td>
          <td>
            9.12 °C
          </td>
          <td>
            10.43 °C
          </td>
          <td>
            10.43 °C
          </td>
        </tr>
      </tbody>
    </table>
  </li>
</ul>
`;

exports[`testing Expand component Matches the snapshot: shows non-expanded Expand component 1`] = `
<ul
  className="detailWeather"
>
  <li>
    <button
      className="expand"
      onClick={[Function]}
    >
      +
    </button>
  </li>
  <li>
    <table>
      <thead>
        <tr>
          <td>
            Date & time
          </td>
          <td>
            Current Weather
          </td>
          <td>
            Current Temp
          </td>
          <td>
            Current Body Temp
          </td>
          <td>
            Current Max Temp
          </td>
          <td>
            Current Min Temp
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            undefined
          </td>
          <td>
            <img
              alt="Algiers 2023-01-18 09:00:00 weather icon"
              className="expand_img"
              src="https://openweathermap.org/img/wn/10d@2x.png"
            />
          </td>
          <td>
            12.01 °C
          </td>
          <td>
            11.12 °C
          </td>
          <td>
            9.79 °C
          </td>
          <td>
            12.01 °C
          </td>
        </tr>
      </tbody>
    </table>
  </li>
</ul>
`;
